addons/project_timesheet_holidays/models/hr_holidays.py

```diff
@@ -151,3 +151,14 @@ def _force_cancel(self, *args, **kwargs):
         timesheets = self.sudo().timesheet_ids
         timesheets.holiday_id = False
         timesheets.unlink()
+
+    def write(self, vals):
+        res = super().write(vals)
+        # reevaluate timesheets after the leaves are wrote in order to remove empty timesheets
+        empty_timesheets = self.env['account.analytic.line'].browse()
+        for leave in self:
+            if leave.number_of_days == 0 and leave.timesheet_ids and leave.timesheet_ids.holiday_id:
```

Comment:
`leave.timesheet_ids.holiday_id` will always be truly since the one2many field `timesheet_ids` will fetch the timesheet_ids with `holiday_id == leave` 🙂 

```suggestion
            if leave.number_of_days == 0 and leave.sudo().timesheet_ids:
```

we need a sudo to be sure to fetch the timesheets even if the current user has no access to timesheets app

See its definition:
https://github.com/odoo/odoo/blob/5eb6054f1071f7fdc4e13982f017abf72089f830/addons/project_timesheet_holidays/models/hr_holidays.py#L56

Reply:
Thanks for the review and the explanations, I applied the changes.